error: splitted error_message document, added error message format

Piankero opened the pull request:
## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [ ] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins and doc/METADATA.md)

## Review

Remove the line below and add the "work in progress" label if you do
not want the PR to be reviewed:

@markus2330 please review my pull request

## Merging

Please add the "ready to merge" label when the build server and you say
that everything is ready to be merged.

Piankero commented:
jenkins build libelektra please

Piankero commented:
@markus2330 does the build server cache README.md documents? The build failure says it cannot find a link when I clearly removed it.

markus2330 commented:
Maybe there is a link to the now non-existing document?

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+# Error message & handling concept
```
Name of the file does not match with this heading.

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+
+## Constraints
+
+- Error numbers must stay because they are more reliable to check against than strings
```
So we keep error numbers? Why is #2686 then changing int -> char*?

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+
+- Error numbers must stay because they are more reliable to check against than strings
+- Supporting multiple programming languages
+- Plugin System
```
```suggestion
- Supporting Elektra's Plugin System
```

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+
+## Decision
+
+All "fatal" errors will be converted to "errors" as the distinguishment is not relevant.
```
```suggestion
All "fatal" errors will be converted to "errors" as the distinction is not relevant.
```

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+
+All "fatal" errors will be converted to "errors" as the distinguishment is not relevant.
+
+Unused marked errors will be removed from the specification.
```
```suggestion
Unused errors will be removed from the specification.
```

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+
+Unused marked errors will be removed from the specification.
+
+Errors will be categorizes into logical groups with subgroups.
```
```suggestion
Errors will be categorized into logical groups with subgroups.
```

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+and character 3 to 5 will be used for subgrouping. Errors are prepended with the letter `C` which
+is the abbreviation for "Code".
+
+- Permanent errors
```
Number missing?

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+The grouping of errors will allow developers to filter for specific as well as more general errors to correctly
+react to them programmatically.
+The new concept will permit additional subgrouping of errors in case it might be needed in the future. Imagine the case where
+"Installation" errors is too general because developers saw a need for splitting the errors in "elektra installation"
```
This is now about the semantics of "Installation errors" but this document should only be about error codes?

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+  - Logical
+    - Assertion C01410
+    - Interface C01420
+    - Broken plugin C01430
```
Misleading name. What of the plugin is broken?

Piankero commented:
```
error: Failed to find link to error_message.md in the file /home/jenkins/workspace/libelektra_PR-2684-I4G4C7O3FAKCIOP7INAZYROCYG3USUUVQMDO2WSLRQGR7T3U2XJA/doc/decisions/README.md

testscr_check_doc.sh RESULTS: 44 test(s) done 1 error(s).
```

There is clearly no error_message.md in the decision/README.md file anymore

markus2330 commented on doc/decisions/error_codes.md:
```diff
@@ -0,0 +1,126 @@
+
+Current errors will be migrated.
+
+## Related Decisions
```
Links to the other decisions is missing. How do they interrelate?

markus2330 commented on doc/decisions/error_message_format.md:
```diff
@@ -0,0 +1,72 @@
+# Error message & handling concept
```
Again the title here does not match the file name (and also not the content).

markus2330 commented on doc/decisions/error_message_format.md:
```diff
@@ -0,0 +1,72 @@
+## Considered Alternatives
+
+Possible variations on what message should be displayed,
+eg. to keep the mountpoint information or on how wordings should be (with or without
```
```suggestion
e.g., to keep the mountpoint information or on how wordings should be (with or without
```

markus2330 commented on doc/decisions/error_message_format.md:
```diff
@@ -0,0 +1,72 @@
+
+Possible variations on what message should be displayed,
+eg. to keep the mountpoint information or on how wordings should be (with or without
+"Sorry, ...", coloring of certain parts of a message, etc.)
```
Make a list of considered alternatives.

Piankero replied:
The alternatives are pretty much endless on how you can rearrange an error message. I added some examples but I think it's is not feasible to list all possible variations of is possible for every field.

markus2330 replied:
The list of alternatives is always endless. It is about listing the *useful* but nevertheless discarded alternatives.

markus2330 commented on doc/decisions/error_message_format.md:
```diff
@@ -0,0 +1,72 @@
+The new error message is much more succinct which gives end users more relevant information.
+Furthermore the solution approach still holds all necessary information if requested by users.
+
+`Ingroup`, `Description` and `Module` will be removed from the error message as they provide no useful
```
This is Implication not Rationale? Description needs to be ported into Reason.

sanssecours commented:
> There is clearly no error_message.md in the decision/README.md file anymore. And the other file exists

The problem here is that the file `error_message.md` still exists, while README.md contains no link to this file. You can fix this error by removing `error_message.md`.

markus2330 commented:
I have already added my review. Please carefully bring the whole docu up-to-date.

Piankero commented:
> I have already added my review. Please carefully bring the whole docu up-to-date.

Beside the alternatives I changed everything you wrote. I know also wrote more alternatives

markus2330 commented:
Thank you.
